pulsar-broker/src/test/java/org/apache/pulsar/client/impl/TableViewTest.java

```diff
@@ -438,4 +444,55 @@ public void testTableViewTailMessageReadRetry() throws Exception {
         });
         verify(consumer, times(msgCnt)).receiveAsync();
     }
+
+    @Test
+    public void testBuildTableViewWithMessagesAlwaysAvailable() throws Exception {
+        String topic = "persistent://public/default/testBuildTableViewWithMessagesAlwaysAvailable";
+        admin.topics().createPartitionedTopic(topic, 10);
+        @Cleanup
+        Reader<byte[]> reader = pulsarClient.newReader()
+                .topic(topic)
+                .startMessageId(MessageId.earliest)
+                .create();
+        @Cleanup
+        Producer<byte[]> producer = pulsarClient.newProducer()
+                .topic(topic)
+                .create();
+        // Prepare real data to do test.
+        for (int i = 0; i < 1000; i++) {
+            producer.newMessage().send();
+        }
+        List<TopicMessageId> lastMessageIds = reader.getLastMessageIds();
+
+        // Use mock reader to build tableview. In the old implementation, the readAllExistingMessages method
+        // will not be completed because the `mockReader.hasMessageAvailable()` always return ture.
+        Reader<byte[]> mockReader = spy(reader);
+        when(mockReader.hasMessageAvailable()).thenReturn(true);
+        when(mockReader.getLastMessageIdsAsync()).thenReturn(CompletableFuture.completedFuture(lastMessageIds));
+        AtomicInteger index = new AtomicInteger(lastMessageIds.size());
+        when(mockReader.readNextAsync()).thenAnswer(invocation -> {
+            Message<byte[]> message = spy(Message.class);
+            int localIndex = index.decrementAndGet();
+            when(message.getTopicName()).thenReturn(lastMessageIds.get(localIndex).getOwnerTopic());
+            when(message.getMessageId()).thenReturn(lastMessageIds.get(localIndex));
+            when(message.hasKey()).thenReturn(false);
+            doNothing().when(message).release();
+            return CompletableFuture.completedFuture(message);
+        });
+        @Cleanup
+        TableViewImpl<byte[]> tableView = (TableViewImpl<byte[]>) pulsarClient.newTableView()
+                .topic(topic)
+                .createAsync()
+                .get();
+        TableViewImpl<byte[]> mockTableView = spy(tableView);
+        Method readAllExistingMessagesMethod = TableViewImpl.class
+                .getDeclaredMethod("readAllExistingMessages", Reader.class);
+        readAllExistingMessagesMethod.setAccessible(true);
+        CompletableFuture<Reader<?>> future =
```

Comment:
Should we send more messages before invoking the `readAllExistingMessages` method to ensure we did not receive the new messages?